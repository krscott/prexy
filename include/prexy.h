#ifndef PREXY_H_
#define PREXY_H_

#ifndef PREXY_EXPAND
#include <assert.h>
#endif

#define prexy

#define prexy_count(name) (name##_COUNT + (0 * sizeof((enum name)0)))

#define PREXY_CONCAT(x, y) x##y
#define PREXY_CONCAT3(x, y, z) x##y##z

#ifdef PREXY_EXPAND

#define PREXY_ASSERT_IS_TAG(name) PREXY_EXPAND_REMOVE_SEMI
#define PREXY_ASSERT_IS_NAMESPACE(name) PREXY_EXPAND_REMOVE_SEMI
#define PREXY_ASSERT_IS_STRUCT(name) PREXY_EXPAND_REMOVE_SEMI

#define prexy_tag(name) PREXY_EXPAND_REMOVE_SEMI
#define prexy_namespace(name) PREXY_EXPAND_REMOVE_SEMI

#define px_tag(name) PREXY_EXPAND_REMOVE_SEMI
#define px_attr(name, ...) PREXY_EXPAND_REMOVE_SEMI

#else

#define PREXY_ASSERT_IS_TAG(name)                                              \
    static_assert(sizeof((struct name){.prexy_tag__ = '\0'}), "")
#define PREXY_ASSERT_IS_NAMESPACE(name)                                        \
    static_assert(sizeof((struct name){.prexy_namespace__ = '\0'}), "")
#define PREXY_ASSERT_IS_STRUCT(name)                                           \
    static_assert(sizeof(*(struct name *)0), "")
// #define PREXY_ASSERT_IS_NONEMPTY_STRING(s)                                     \
//     static_assert(sizeof("" s "") > 1, "")

#define prexy_tag(name)                                                        \
    struct name                                                                \
    {                                                                          \
        char prexy_tag__;                                                      \
    }
#define prexy_namespace(name)                                                  \
    struct name                                                                \
    {                                                                          \
        char prexy_namespace__;                                                \
    }

#define px_tag(name) PREXY_ASSERT_IS_TAG(name)
#define px_attr(name, ...) static_assert(sizeof((struct name){__VA_ARGS__}), "")

#endif

// Pre-processor marker macros

// Struct function decl/impl macros

#define prexy_decl(name, macro_name)                                           \
    macro_name##_decl(name);                                                   \
    PREXY_ASSERT_IS_NAMESPACE(macro_name)

// clang-format off
// NOTE: _impl macros do not end with semicolon

#define prexy_impl(name, macro_name)                                           \
    macro_name##_impl(name, name##_X)                                          \
    PREXY_ASSERT_IS_NAMESPACE(macro_name)

#define prexy_impl_attr(name, macro_name, attr)                                \
    macro_name##_impl(name, name##_X_##attr)                                   \
    PREXY_ASSERT_IS_STRUCT(attr)

#define prexy_impl_tag(name, macro_name, tag)                                  \
    macro_name##_impl(name, name##_X_##tag)                                    \
    PREXY_ASSERT_IS_NAMESPACE(macro_name)

// clang-format on

// Generic macros

#define prexy_typeof(struct_name, field_name)                                  \
    PREXY_CONCAT3(struct_name, _FIELDTYPE_, field_name)
#define prexy_is_mut_ptr(struct_name, field_name)                              \
    PREXY_CONCAT3(struct_name, _IS_MUT_PTR_, field_name)
#define prexy_is_const_ptr(struct_name, field_name)                            \
    PREXY_CONCAT3(struct_name, _IS_CONST_PTR_, field_name)
#define prexy_ptr_typeof(struct_name, field_name)                              \
    PREXY_CONCAT3(struct_name, _PTRTYPE_, field_name)

#define prexy_methodname(struct_name, fn_name)                                 \
    PREXY_CONCAT3(struct_name, _, fn_name)

#endif
