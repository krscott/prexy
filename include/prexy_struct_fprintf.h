#ifndef PREXY_STRUCT_FPRINTF_H_
#define PREXY_STRUCT_FPRINTF_H_

#include <assert.h>
#include <stdio.h>

// Get printf format string for a type.
// Does *not* assume a `char *` is a `%s`.
#define PREXY_UTIL_SIMPLE_TYPE_FORMAT(x)                                       \
    (_Generic(                                                                 \
        (x),                                                                   \
         _Bool: "%d",                                                          \
         short: "%hd",                                                         \
         unsigned short: "%hu",                                                \
         int: "%d",                                                            \
         unsigned int: "%u",                                                   \
         long: "%ld",                                                          \
         unsigned long: "%lu",                                                 \
         long long: "%lld",                                                    \
         unsigned long long: "%llu",                                           \
         float: "%f",                                                          \
         double: "%f",                                                         \
         char: "'%c'",                                                         \
         void *: "%p"                                                          \
    ))

#define PREXY_STRUCT_FPRINT_REPR_simple(type, varname)                         \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        fprintf(                                                               \
            stream,                                                            \
            PREXY_UTIL_SIMPLE_TYPE_FORMAT((x)->varname),                       \
            (x)->varname                                                       \
        );                                                                     \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_simple_array(type, varname, len)              \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            fprintf(                                                           \
                stream,                                                        \
                PREXY_UTIL_SIMPLE_TYPE_FORMAT((x)->varname[i]),                \
                (x)->varname[i]                                                \
            );                                                                 \
            fprintf(stream, ", ");                                             \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_struct(type, varname)                         \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        type##_fprint_repr(stream, &(x)->varname);                             \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_struct_array(type, varname, len)              \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            type##_fprint_repr(stream, &(x)->varname[i]);                      \
            fprintf(stream, ", ");                                             \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_enum(type, varname)                           \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        fprintf(stream, "%s", type##_to_cstr((x)->varname));                   \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_enum_array(type, varname, len)                \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            fprintf(stream, "%s, ", type##_to_cstr((x)->varname[i]));          \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR(fkind, ...)                                   \
    PREXY_STRUCT_FPRINT_REPR_##fkind(__VA_ARGS__)

#define prexy_struct_fprint_repr_decl(name)                                    \
    void name##_fprint_repr(FILE *stream, struct name const *x)
#define prexy_struct_fprint_repr_impl_x(name, FIELDS_X)                        \
    prexy_struct_fprint_repr_decl(name)                                        \
    {                                                                          \
        fprintf(stream, "(struct " #name "){ ");                               \
        FIELDS_X(PREXY_STRUCT_FPRINT_REPR)                                     \
        fprintf(stream, "}");                                                  \
    }                                                                          \
    static_assert(1, "")
#define prexy_struct_fprint_repr_impl(name)                                    \
    prexy_struct_fprint_repr_impl_x(name, name##_x_fields)
#define prexy_struct_fprint_repr_impl_attr(name, attr)                         \
    prexy_struct_fprint_repr_impl_x(name, name##_x_fields_##attr);             \
    static_assert(sizeof(*(struct attr *)0), "")

#endif
