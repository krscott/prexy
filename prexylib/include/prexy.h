#ifndef PREXY_H_
#define PREXY_H_

#include <assert.h>

// Pre-processor marker macro

#define prexy
#define prexy_tag(name)                                                        \
    struct name                                                                \
    {                                                                          \
        char prexy_tag__;                                                      \
    }

#define PREXY_ASSERT_IS_TAG(name)                                              \
    static_assert(sizeof(((struct name *)0)->prexy_tag__), "")

#define px_tag(name) static_assert(sizeof((struct name){'\0'}), "")
#define px_attr(name, ...) static_assert(sizeof((struct name){__VA_ARGS__}), "")

// Struct function decl/impl macros

#define prexy_decl(name, macro_name)                                           \
    macro_name##_decl(name);                                                   \
    PREXY_ASSERT_IS_TAG(macro_name)

// clang-format off
// NOTE: _impl macros do not end with semicolon

#define prexy_enum_impl(name, macro_name)                                      \
    macro_name##_impl(name, name##_x_variants)                                 \
    PREXY_ASSERT_IS_TAG(macro_name)

#define prexy_impl(name, macro_name)                                           \
    macro_name##_impl(name, name##_x_fields)                                   \
    PREXY_ASSERT_IS_TAG(macro_name)

#define prexy_impl_attr(name, macro_name, attr)                                \
    macro_name##_impl(name, name##_x_fields_##attr)                            \
    static_assert(sizeof(*(struct attr *)0), "")

#define prexy_impl_tag(name, macro_name, tag)                                  \
    macro_name##_impl(name, name##_x_fields_##tag)                             \
    PREXY_ASSERT_IS_TAG(tag)

// clang-format on

#endif
