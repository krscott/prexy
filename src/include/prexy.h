#ifndef PREXY_H_
#define PREXY_H_

// Markers
#define prexy

#endif
