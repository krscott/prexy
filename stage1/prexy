#!/usr/bin/env bash
set -euo pipefail
shopt -s failglob

DIR="$(dirname "$(realpath "${BASH_SOURCE[0]}")")"

usage() {
    # NOTE: Don't forget to update OPTIONS/LONGOPTS/POS_MIN/POS_MAX below!
    cat <<EOF
Usage: $(basename "$0") [options] source

Positional Parameters:
    source                  Source C file to preprocess

Options:
    --header                Print prexy.h
    -o, --output PATH       Output file path
    -f, --formatter CMD     Formatter command (default: 'clang-format')
    -E, --expand            Run expand-preprocessor codegen
    -I, --include           Add include directory (used with -E)
    -h, --help              Display this help message.
    -v, --verbose           Enable debug messages
EOF
}

# NOTE: Use ':' suffix to specify opt has an arg
OPTIONS="o:f:EI:hv"
LONGOPTS="header,output:,formatter:,expand,include:,help,verbose"
POS_MIN=1
POS_MAX=1

if ! PARSED=$(getopt --options="$OPTIONS" --longoptions="$LONGOPTS" --name "$0" -- "$@"); then
    usage
    exit 2
fi

eval set -- "$PARSED"

print_header=
output=
formatter=clang-format
expand=
include_args=(".")
include_flags=()
verbose=

while true; do
    opt="$1"
    shift

    # NOTE: shift again if taking an arg

    case "$opt" in
    --header)
        print_header=1
        ;;
    -o | --output)
        output="$1"
        shift
        ;;
    -f | --formatter)
        formatter="$1"
        shift
        ;;
    -E | --expand)
        expand=1
        ;;
    -I | --include)
        include_args+=("$1")
        include_flags+=("-I$1")
        shift
        ;;
    -h | --help)
        usage
        exit 0
        ;;
    -v | --verbose)
        verbose=1
        ;;
    --)
        break
        ;;
    *)
        echo "Unknown option: $opt" >&2
        usage
        exit 2
        ;;
    esac
done

if [[ -n $print_header ]]; then
    cat "$DIR/../include/prexy.h"
    exit 0
fi

if [[ $# -lt $POS_MIN ]]; then
    echo "Missing argument" >&2
    usage
    exit 2
fi
if [[ $# -gt $POS_MAX ]]; then
    echo "Too many arguments" >&2
    usage
    exit 2
fi

source_path="$1"

if [[ ! -f $source_path ]]; then
    echo "File '$source_path' does not exist"
    exit 1
fi

debug() {
    if [[ -n $verbose ]]; then
        echo "$@" >&2
    fi
}

verbose() {
    if [[ -n $verbose ]]; then
        set -x
    fi
}

find_include() {
    for dir in "${include_args[@]}"; do
        if [[ -f "$dir/$1" ]]; then
            realpath "$dir/$1"
            return
        fi
    done

    echo "Could not find file: $1" >&2
    exit 3
}

find_expanded_includes() {
    local input
    input=$(find_include "$1")

    local includes
    mapfile -t includes < <(awk -f "$DIR/find_expand_includes.awk" "$input")
    for include in "${includes[@]}"; do
        find_expanded_includes "$include"
    done
    echo "$input"
}

uniq_unsorted() {
    awk '!seen[$0]++'
}

codegen() {
    verbose

    if [[ -n $expand ]]; then
        mapfile -t sources < <(find_expanded_includes "$source_path" | uniq_unsorted)

        for src in "${sources[@]}"; do
            awk -f "$DIR/unexpand.awk" "$src"
        done

        echo ""

        $CC -E -P "${include_flags[@]}" -DPREXY_EXPAND "$source_path"
    else
        debug "Formatting '$source_path'"
        awk -f "$DIR/prexy.awk" "$source_path"
    fi
}

if [[ -n $output ]]; then
    codegen | $formatter >"$output"
else
    codegen | $formatter
fi
