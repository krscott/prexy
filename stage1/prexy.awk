#!/usr/bin/env -S awk -f

BEGIN {
    IDLE = 0
    COLLECT = 1

    state = IDLE
}

FNR==1 {
    fname = FILENAME
    sub(".*/", "", fname)  # remove Unix-style path
    sub(".*\\\\", "", fname)  # remove Windows-style path

    fname_ident = fname
    sub("\\.[^.]*$", "", fname_ident)  # remove extension
    gsub(/[^[:alnum:]]+/, "_", fname_ident)
    sub(/^_+/, "", fname_ident)
    sub(/_+$/, "", fname_ident)
    fname_ident = toupper(fname_ident)
    print "#ifndef PREXY_CLIENT_" fname_ident "_H_"
    print "#define PREXY_CLIENT_" fname_ident "_H_"
    print ""
    print "/* Generated by prexy from: " fname " */"
    print ""
    print "#include \"prexy.h\""
    print ""
}

END {
    print "#endif"
}

/^prexy\s/ {
    if (state != IDLE) {
        print "// " $0
        error("Unexpected 'prexy' while parsing other structure")
    }
    state = COLLECT
    code = ""
}
state == COLLECT {
    print "// " $0
    gsub(/\/\/.*$/, "", $0) # Remove comments
    code = code $0
}
state == COLLECT && /};/ {
    state = IDLE

    code = strip(code)

    split("", data)  # clear data array

    if (parse_enum(code, data)) {
        print_enum(data)
    } else if (parse_struct(code, data)) {
        print_struct(data)
    } else {
        error("Unknown structure type")
    }

    print ""
}

function parse_enum(code, data) {
    if (!match(code, /^prexy enum ([[:alnum:]_]+)[[:space:]]*{(.*)};$/, m)) {
        return 0
    }

    data["name"] = m[1]

    count = 0
    split(m[2], variants, ",")
    for (i in variants) {
        if (match(strip(variants[i]), /^([[:alnum:]_]+)/, m)) {
            data["variants"][count++] = m[1]
        }
    }

    data["count"] = count

    return 1
}

function print_enum(data) {
    print "#define " data["name"] "_x_count " data["count"]
    print "#define " data["name"] "_x_variants(X) \\"
    for (i = 0; i < data["count"]; ++i) {
        print "X(" data["variants"][i] ") \\"
    }
    print ""
}

function parse_struct(code, data) {
    if (!match(code, /^prexy struct ([[:alnum:]_]+)[[:space:]]*{(.*)};$/, m)) {
        print code
        return 0
    }

    data["name"] = m[1]

    # Clear arrays
    split("", data["attrs"])
    split("", data["fields"])
    split("", attr_exists)

    field_count = 0
    attr_count = 1  # attr 0 is implicit "no-attributes" attr

    split(m[2], fields, ";")
    for (i in fields) {
        s = strip(fields[i])

        data["fields"][field_count]["length"] = 0

        if (match(s, /^px_tag\([[:space:]]*([[:alnum:]_]+)[[:space:]]*\)$/, m)) {
            # tag (0-dimensional attribute)
            if (!attr_exists[m[1]]) {
                attr_exists[m[1]] = 1
                data["attrs"][attr_count] = m[1]
                ++attr_count
            }
            data["fields"][field_count]["tags"][m[1]] = 1
        } else if (match(s, /^px_attr\([[:space:]]*([[:alnum:]_]+),[[:space:]]+(.*)\)$/, m)) {
            # attribute
            if (!attr_exists[m[1]]) {
                attr_exists[m[1]] = 1
                data["attrs"][attr_count] = m[1]
                ++attr_count
            }
            data["fields"][field_count]["attrs"][m[1]] = m[2]
        } else if (match(s, /^(enum|struct)[[:space:]]+([[:alnum:]_]+)[[:space:]+]([[:alnum:]_]+)\[[[:space:]]*([[:alnum:]]+)[[:space:]]*\]$/, m)) {
            # array of enums or structs
            data["fields"][field_count]["metatype"] = m[1] "_array"
            data["fields"][field_count]["type"] = strip(m[2])
            data["fields"][field_count]["name"] = strip(m[3])
            data["fields"][field_count]["length"] = strip(m[4])
            ++field_count
        } else if (match(s, /^(enum|struct)[[:space:]]+([[:alnum:]_]+)[[:space:]+]([[:alnum:]_]+)$/, m)) {
            # enum or struct
            data["fields"][field_count]["metatype"] = m[1]
            data["fields"][field_count]["type"] = strip(m[2])
            data["fields"][field_count]["name"] = strip(m[3])
            ++field_count
        } else if (match(s, /^([[:alnum:]_*[:space:]]+[[:space:]*])([[:alnum:]_]+)\[[[:space:]]*([[:alnum:]]+)[[:space:]]*\]$/, m)) {
            # array of simple types
            data["fields"][field_count]["metatype"] = "simple_array"
            data["fields"][field_count]["type"] = strip(m[1])
            data["fields"][field_count]["name"] = strip(m[2])
            data["fields"][field_count]["length"] = strip(m[3])
            ++field_count
        } else if (match(s, /^([[:alnum:]_*[:space:]]+[[:space:]*])([[:alnum:]_]+)$/, m)) {
            # simple type
            data["fields"][field_count]["metatype"] = "simple"
            data["fields"][field_count]["type"] = strip(m[1])
            data["fields"][field_count]["name"] = strip(m[2])
            ++field_count
        }
    }

    data["field_count"] = field_count
    data["attr_count"] = attr_count

    return 1
}

function print_struct(data) {
    for (i = 0; i < data["attr_count"]; ++i) {
        attr_name = data["attrs"][i]

        if (i == 0) {
            # attr 0 is implicit "no-attributes" attr
            print "#define " data["name"] "_x_fields(F) \\"
        } else {
            print "#define " data["name"] "_x_fields_" attr_name "(F) \\"
        }

        for (j = 0; j < data["field_count"]; ++j) {
            field_attr = data["fields"][j]["attrs"][attr_name]
            field_tag = data["fields"][j]["tags"][attr_name]
            printf "F("

            # attribute/tag name or metatype
            if (field_attr != "" || field_tag != "") {
                printf attr_name
            } else {
                printf data["fields"][j]["metatype"]
            }

            # type and field name
            printf ", " data["fields"][j]["type"] ", " data["fields"][j]["name"]

            # array length
            if (data["fields"][j]["length"]) {
                printf ", " data["fields"][j]["length"]
            }

            # attribute struct fields
            if (field_attr != "") {
                printf ", " data["fields"][j]["attrs"][attr_name]
            }

            print ") \\"
        }
        print ""
    }
}

function error(msg) {
    print "error: " msg > "/dev/stderr"
    print "#error \"prexy: " msg "\""
    exit 1
}

function strip(s) {
    gsub(/^[[:space:]]+/, "", s)
    gsub(/[[:space:]]+$/, "", s)
    return s
}

