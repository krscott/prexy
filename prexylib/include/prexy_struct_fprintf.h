#ifndef PREXY_STRUCT_FPRINTF_H_
#define PREXY_STRUCT_FPRINTF_H_

#include "prexy.h"
#include <assert.h>
#include <stdio.h>

// Get printf format string for a type.
// Does *not* assume a `char *` is a `%s`.
#define PREXY_UTIL_SIMPLE_TYPE_FORMAT(x)                                       \
    (_Generic(                                                                 \
        (x),                                                                   \
         _Bool: "%d",                                                          \
         short: "%hd",                                                         \
         unsigned short: "%hu",                                                \
         int: "%d",                                                            \
         unsigned int: "%u",                                                   \
         long: "%ld",                                                          \
         unsigned long: "%lu",                                                 \
         long long: "%lld",                                                    \
         unsigned long long: "%llu",                                           \
         float: "%f",                                                          \
         double: "%f",                                                         \
         char: "'%c'",                                                         \
         void *: "%p"                                                          \
    ))

#define PREXY_STRUCT_FPRINT_REPR_simple(type, varname)                         \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        fprintf(                                                               \
            stream,                                                            \
            PREXY_UTIL_SIMPLE_TYPE_FORMAT((x)->varname),                       \
            (x)->varname                                                       \
        );                                                                     \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_simple_array(type, varname, len)              \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            fprintf(                                                           \
                stream,                                                        \
                PREXY_UTIL_SIMPLE_TYPE_FORMAT((x)->varname[i]),                \
                (x)->varname[i]                                                \
            );                                                                 \
            fprintf(stream, ", ");                                             \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_struct(type, varname)                         \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        type##_fprint_repr(stream, &(x)->varname);                             \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_struct_array(type, varname, len)              \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            type##_fprint_repr(stream, &(x)->varname[i]);                      \
            fprintf(stream, ", ");                                             \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_enum(type, varname)                           \
    {                                                                          \
        fprintf(stream, "." #varname " = ");                                   \
        fprintf(stream, "%s", type##_to_cstr((x)->varname));                   \
        fprintf(stream, ", ");                                                 \
    }

#define PREXY_STRUCT_FPRINT_REPR_enum_array(type, varname, len)                \
    {                                                                          \
        fprintf(stream, "." #varname " = { ");                                 \
        for (int i = 0; i < len; ++i)                                          \
        {                                                                      \
            fprintf(stream, "%s, ", type##_to_cstr((x)->varname[i]));          \
        }                                                                      \
        fprintf(stream, "}, ");                                                \
    }

#define PREXY_STRUCT_FPRINT_REPR_FIELD_SELECT(fkind, ...)                      \
    PREXY_STRUCT_FPRINT_REPR_##fkind(__VA_ARGS__)

prexy_namespace(prexy_struct_fprint_repr);

#define prexy_struct_fprint_repr_decl(name)                                    \
    void name##_fprint_repr(FILE *stream, struct name const *x)

#define prexy_struct_fprint_repr_impl(name, FIELDS_X)                          \
    prexy_struct_fprint_repr_decl(name)                                        \
    {                                                                          \
        fprintf(stream, "(struct " #name "){ ");                               \
        FIELDS_X(PREXY_STRUCT_FPRINT_REPR_FIELD_SELECT)                        \
        fprintf(stream, "}");                                                  \
    }

#endif
