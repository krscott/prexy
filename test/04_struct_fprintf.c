#include "04_struct_fprintf_prexy.h"
#include "prexy.h"
#include "prexy_struct_fprintf.h"
#include <stdbool.h>
#include <stdio.h>

prexy enum cheese_kind {
    CHEDDAR,
    SWISS,
    BRIE,
};
static prexy_enum_to_cstr_impl(cheese_kind);

prexy struct cheese
{
    enum cheese_kind kind;
    bool is_melty;
};
static prexy_struct_fprint_repr_impl(cheese);

prexy struct bagel
{
    char name[10];
    double price;
    struct cheese cheeses[3];
};
static prexy_struct_fprint_repr_impl(bagel);

int main(void)
{
    struct bagel bagel = {
        .name = "everything",
        .price = 5.99,
        .cheeses =
            {
                {.kind = CHEDDAR, .is_melty = true},
                {.kind = SWISS, .is_melty = false},
                {.kind = BRIE, .is_melty = false},
            },
    };

    bagel_fprint_repr(stderr, &bagel);
}
