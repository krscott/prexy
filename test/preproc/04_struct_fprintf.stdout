(struct bagel){ .name = { 'e', 'v', 'e', 'r', 'y', 't', 'h', 'i', 'n', 'g', }, .price = 1.990000, .cheeses = { (struct cheese){ .kind = CHEDDAR, .is_melty = 1, }, (struct cheese){ .kind = SWISS, .is_melty = 0, }, (struct cheese){ .kind = BRIE, .is_melty = 0, }, }, }